rust/noosphere-storage/src/implementation/performance.rs

```diff
@@ -0,0 +1,211 @@
+pub struct InternalStoreStats {
+    pub reads: usize,
+    pub writes: usize,
+    pub removes: usize,
+    pub flushes: usize,
+    pub reads_duration_us: u128,
+    pub writes_duration_us: u128,
+    pub removes_duration_us: u128,
+    pub flushes_duration_us: u128,
+    pub logical_bytes_stored: u64,
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct PerformanceStats {
+    pub reads: usize,
+    pub writes: usize,
+    pub removes: usize,
+    pub flushes: usize,
+    pub median_reads_us: u128,
+    pub median_writes_us: u128,
+    pub median_removes_us: u128,
+    pub median_flushes_us: u128,
+    pub logical_bytes_stored: u64,
+    pub physical_bytes_stored: u64,
+}
+
+#[derive(Clone, Debug)]
+pub struct PerformanceStorage<S: Storage> {
+    storage: S,
+    stats: Arc<Mutex<HashMap<String, Arc<Mutex<InternalStoreStats>>>>>,
+}
+
+impl<S> PerformanceStorage<S>
+where
+    S: Storage,
+    S::KeyValueStore: Store,
+    S::BlockStore: Store,
+{
+    pub fn new(other: S) -> Self {
+        PerformanceStorage {
+            storage: other,
+            stats: Arc::new(Mutex::new(HashMap::default())),
+        }
+    }
+
+    async fn get_store_stats(&self, name: &str) -> Arc<Mutex<InternalStoreStats>> {
+        let mut storage_stats = self.stats.lock().await;
+        let store_name = name.to_owned();
+        if let Some(stats) = storage_stats.get(&store_name) {
+            stats.to_owned()
+        } else {
+            let store_stats = Arc::new(Mutex::new(InternalStoreStats::default()));
+            storage_stats.insert(store_name, store_stats.clone());
+            store_stats
+        }
+    }
+
+    pub fn to_inner(self) -> S {
+        self.storage
+    }
+}
+
+impl<S> PerformanceStorage<S>
+where
+    S: Storage + Space,
+    S::KeyValueStore: Store,
+    S::BlockStore: Store,
+{
+    pub async fn to_stats(&self) -> Result<PerformanceStats> {
+        let storage_stats = self.stats.lock().await;
+        let mut agg = InternalStoreStats::default();
+        for (_, store_stats) in storage_stats.iter() {
+            let stats = store_stats.lock().await;
+            agg.reads += stats.reads;
+            agg.writes += stats.writes;
+            agg.removes += stats.removes;
+            agg.flushes += stats.flushes;
+            agg.reads_duration_us += stats.reads_duration_us;
+            agg.writes_duration_us += stats.writes_duration_us;
+            agg.removes_duration_us += stats.removes_duration_us;
+            agg.flushes_duration_us += stats.flushes_duration_us;
+            agg.logical_bytes_stored += stats.logical_bytes_stored;
+        }
+
+        let mut stats = PerformanceStats::default();
+        stats.reads = agg.reads;
+        stats.writes = agg.writes;
+        stats.removes = agg.removes;
+        stats.flushes = agg.flushes;
+        stats.median_reads_us = if agg.reads == 0 {
+            0
+        } else {
+            agg.reads_duration_us / agg.reads as u128
```

Comment:
Is it acceptable to be doing integer division here? Don't you want floating point precision (and rounding as a formatting detail at the end)?